pkg/printer/printer.go

```diff
@@ -384,6 +384,139 @@ func (p *Printer) WriteProtoNodeStatusEvent(r *observerpb.GetFlowsResponse) erro
+				strings.Join(p.Labels, ","), p.Revision, p.RuleCount)
+		}
+	case pb.AgentEventType_ENDPOINT_REGENERATE_SUCCESS, pb.AgentEventType_ENDPOINT_REGENERATE_FAILURE:
+		if r := e.GetEndpointRegenerate(); r != nil {
+			str := fmt.Sprintf("id: %d, labels: [%s]", r.Id, strings.Join(r.Labels, ","))
+			if re := r.Error; re != "" {
+				str += fmt.Sprintf(", error: %s", re)
+			}
+			return str
+		}
+	case pb.AgentEventType_ENDPOINT_CREATED, pb.AgentEventType_ENDPOINT_DELETED:
+		if ep := e.GetEndpointUpdate(); ep != nil {
+			str := fmt.Sprintf("id: %d", ep.Id)
+			if n := ep.PodName; n != "" {
+				str += fmt.Sprintf(", pod name: %s", n)
+			}
+			if n := ep.Namespace; n != "" {
+				str += fmt.Sprintf(", namespace: %s", n)
+			}
+			return str
+		}
+	case pb.AgentEventType_IPCACHE_UPSERTED, pb.AgentEventType_IPCACHE_DELETED:
+		if i := e.GetIpcacheUpdate(); i != nil {
+			str := fmt.Sprintf("cidr: %s, identity: %d", i.Cidr, i.Identity)
+			if i.OldIdentity != nil {
+				str += fmt.Sprintf(", old identity: %d", i.OldIdentity.Value)
+			}
+			if i.HostIp != "" {
+				str += fmt.Sprintf(", host ip: %s", i.HostIp)
+			}
+			if i.OldHostIp != "" {
+				str += fmt.Sprintf(", old host ip: %s", i.OldHostIp)
+			}
+			return str + fmt.Sprintf(", encrypt key: %d", i.EncryptKey)
+		}
+	case pb.AgentEventType_SERVICE_UPSERTED:
+		if svc := e.GetServiceUpsert(); svc != nil {
+			str := fmt.Sprintf("id: %d", svc.Id)
+			if fe := svc.FrontendAddress; fe != nil {
+				str += fmt.Sprintf(", frontend: %s", formatServiceAddr(fe))
+			}
+			if bes := svc.BackendAddresses; len(bes) != 0 {
+				backends := make([]string, 0, len(bes))
+				for _, a := range bes {
+					backends = append(backends, formatServiceAddr(a))
+				}
+				str += fmt.Sprintf(", backends: [%s]", strings.Join(backends, ","))
+			}
+			if t := svc.Type; t != "" {
+				str += fmt.Sprintf(", type: %s", t)
+			}
+			if tp := svc.TrafficPolicy; tp != "" {
+				str += fmt.Sprintf(", traffic policy: %s", tp)
+			}
+			if n := svc.Name; n != "" {
+				str += fmt.Sprintf(", name: %s", n)
+			}
+			if ns := svc.Namespace; ns != "" {
+				str += fmt.Sprintf(", namespace: %s", ns)
+			}
+			return str
+		}
+	case pb.AgentEventType_SERVICE_DELETED:
+		if s := e.GetServiceDelete(); s != nil {
+			return fmt.Sprintf("id: %d", s.Id)
+		}
+	}
+	return "UNKNOWN"
+}
+
+// WriteProtoAgentEvent writes v1.AgentEvent into the output writer.
+func (p *Printer) WriteProtoAgentEvent(r *observerpb.GetFlowsResponse) error {
+	e := r.GetAgentEvent()
+	if e == nil {
+		return errors.New("not an agent event")
+	}
+
+	switch p.opts.output {
+	case JSONOutput, JSONPBOutput:
+		return json.NewEncoder(p.opts.werr).Encode(r)
+	case DictOutput:
+		// TODO
+	case TabOutput:
+		ew := &errWriter{w: p.tw}
+		if p.line == 0 {
+			ew.write(
+				"TIMESTAMP", tab,
+				"NODE", tab,
+				"TYPE", tab,
+				"DETAILS", newline,
+			)
+			p.line++
+		}
+		ew.write(fmtTimestamp(
+			r.GetTime()), tab,
+			r.GetNodeName(), tab,
```

Comment:
I also realized that for the other event types we only write out the node name when `p.opts.nodeName` is set. Will adjust this accordingly as well.